File: 07/Java/TripleExpression/CheckedAdd.java
package TripleExpression;

public class CheckedAdd extends BinaryOperator implements TripleExpression {
    public CheckedAdd(TripleExpression first, TripleExpression second) {
        super(first, second);
    }

    protected void check(int a, int b) throws Exception {
        if (a > 0 && b > Integer.MAX_VALUE - a) {
            throw new OverflowException();    
        }
        if (a < 0 &&  b < Integer.MIN_VALUE - a) {
            throw new OverflowException();
        }
    }

    protected int apply(int a, int b) throws Exception {
        check(a, b);
        return a + b;
    }
}

File: 07/CSharp/TripleExpressionApp/TripleExpressionApp/CheckedAdd.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    public class CheckedAdd : BinaryOperator
    {
        public CheckedAdd(ITripleExpression first, ITripleExpression second)
        {
            fOp = first;
            sOp = second;
        }
        
        protected void Check(double a, double b)
        {
            if (a > 0 && b > double.MaxValue - a)
            {
                throw new OverflowException();
            }
            if (a < 0 && b < double.MinValue - a)
            {
                throw new OverflowException();
            }
        }

        protected override double Apply(double a, double b)
        {
            Check(a, b);
            return a + b;
        }
    }
}
